### PR TITLE
Qwen 2.5 VL

### PR DESCRIPTION
**Pull request by albert-inflection**
#### Context
What is the purpose of this PR? Is it to
- [ X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
https://github.com/pytorch/torchtune/issues/2699

#### Changelog
What are the changes made in this PR?
*custom modules for Qwen 2.5 VL
*model + component builders for all variants
*Transform + custom collation
*Weight loading
*Added unit tests

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

so far we've done
* Logit comparison against HF for encoder, decoder, and combined models (combined shown)
Screenshot 2025-07-03 at 4 00 26 PM

* successful E2E training runs for all model variants (7B run shown)
Screenshot 2025-07-02 at 5 53 53 PM

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [X] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [X] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/modules/attention.py [hunk 1]
```diff
@@ -185,6 +185,7 @@ def forward(
         *,
         mask: Optional[_MaskType] = None,
         input_pos: Optional[torch.Tensor] = None,
+        window_index: Optional[torch.Tensor] = None,
```
Comment (albert-inflection) [inline torchtune/modules/attention.py:219]: just a placeholder - a main dilemma we've had was where the unique window and rope index calculations should live in the stack. Currently, we've tried our best to preserve primitives and patterns but clearly not perfect, would appreciate outside input

File: torchtune/models/qwen2_5_vision/_fusion.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
+from typing import Any, Dict, Optional, Tuple, Union, List
+import torch
+from torch import nn
+from torchtune.modules.model_fusion._early_fusion import EarlyFusionModel
+from torchtune.modules import TransformerDecoder
+
+class Qwen25VL(EarlyFusionModel):
```
Comment (albert-inflection) [inline torchtune/models/qwen2_5_vision/_fusion.py:15]: see https://github.com/pytorch/torchtune/pull/2868/files#r2183949447

**Comment by albert-inflection** [comment]
a bit rocky, early looks appreciated @joecummings

File: torchtune/modules/attention.py [hunk 3]
```diff
@@ -15,6 +16,36 @@
...
+
+    # Only add parameters that the method accepts
+    if "input_pos" in sig.parameters:
+        kwargs["input_pos"] = input_pos
+    if "window_index" in sig.parameters:
+        kwargs["window_index"] = window_index
+
+    return pos_embedding(x, **kwargs)
+
+
```
Comment (lawrence-inflection) [inline torchtune/modules/attention.py:48]: Current workaround for passing window_index into the positional embedding module.
